v1/plugins/logs/plugin_test.go

```diff
@@ -1670,53 +1683,203 @@ func TestPluginRateLimitDropCountStatus(t *testing.T) {
 func TestChunkMaxUploadSizeLimitNDBCacheDropping(t *testing.T) {
 	t.Parallel()
 
-	ctx := context.Background()
-	testLogger := test.New()
-
-	ts, err := time.Parse(time.RFC3339Nano, "2018-01-01T12:00:00.123456Z")
-	if err != nil {
-		panic(err)
+	// note this only tests the size buffer type so that the encoder is used immediately on log
+	tests := []struct {
+		name                             string
+		uploadSizeLimitBytes             int64
+		numOfEventsWithNDCache           int
+		expectedDroppedNDCacheEvents     int
+		numOfEventsWithoutNDCache        int
+		expectedUncompressedBytesWritten int
+		expectedEventsInChunk            []int
+		expectedChunksLengths            []int // this will show the adaptive uncompressed limit adjusting
+		expectedUncompressedLimit        int64
+	}{
+		{
+			name:                         "drop ND cache to fit",
+			uploadSizeLimitBytes:         220,
+			numOfEventsWithNDCache:       1,
+			expectedDroppedNDCacheEvents: 1,
+			// opening bracket (1 byte) + event(214 bytes)
+			// written after dropping the ND cache, otherwise the size would have been 3472
+			expectedUncompressedBytesWritten: 215,
+			expectedUncompressedLimit:        220,
+		},
+		{
+			name:                         "drop ND cache but keep stabilizing the adaptive uncompressed limit",
+			uploadSizeLimitBytes:         220,
+			numOfEventsWithNDCache:       50,
+			expectedDroppedNDCacheEvents: 50,
+			// the adaptive uncompressed limit increases until it stabilizes
+			expectedChunksLengths:            []int{197, 197, 197, 197, 214, 214, 214, 214, 214},
+			expectedEventsInChunk:            []int{1, 1, 1, 1, 9, 9, 9, 9, 9}, // 49 in total, plus one in the encoder
```

Comment:
What am I missing here? Why do we expect this spread of events, and why does a one-event chunk require `197` bytes while a nine-event chunk requires only slightly more (`214`)? Is this an artifact of how the test is setup or of gzip?

Reply:
Should the uncompressed soft-limit be similarly traced and asserted as it changes?